ref/Microsoft.Build.Framework/net/Microsoft.Build.Framework.cs

```diff
@@ -16,11 +16,13 @@ public partial class BuildErrorEventArgs : Microsoft.Build.Framework.LazyFormatt
         public BuildErrorEventArgs(string subcategory, string code, string file, int lineNumber, int columnNumber, int endLineNumber, int endColumnNumber, string message, string helpKeyword, string senderName) { }
         public BuildErrorEventArgs(string subcategory, string code, string file, int lineNumber, int columnNumber, int endLineNumber, int endColumnNumber, string message, string helpKeyword, string senderName, System.DateTime eventTimestamp) { }
         public BuildErrorEventArgs(string subcategory, string code, string file, int lineNumber, int columnNumber, int endLineNumber, int endColumnNumber, string message, string helpKeyword, string senderName, System.DateTime eventTimestamp, params object[] messageArgs) { }
+        public BuildErrorEventArgs(string subcategory, string code, string file, int lineNumber, int columnNumber, int endLineNumber, int endColumnNumber, string message, string helpKeyword, string helpLink, string senderName, System.DateTime eventTimestamp, params object[] messageArgs) { }
```

Comment:
I'm torn between the current approach (which is most logical IMO) and this, which would be more likely to keep existing code working if someone adds the link. Anyone else have an opinion?

```suggestion
        public BuildErrorEventArgs(string subcategory, string code, string file, int lineNumber, int columnNumber, int endLineNumber, int endColumnNumber, string message, string helpKeyword, string senderName, string helpLink, System.DateTime eventTimestamp, params object[] messageArgs) { }
```

Reply:
I'm not sure I understand why flipping the order changes whether something will keep working. For either case, someone would add a new parameter, just before the senderName or after it, so the user would have to make sure they get it right. Unless I'm misunderstanding the problem, I'd have a weak preference for keeping helpKeyword and helpLink together.

Reply:
I think I'm with @Forgind on this. I don't see much value in swapping the arguments, especially since they are both `string`'s. As a user of this API you need to be aware of the parameter order anyway.

Reply:
Putting API review hat on, @rainersigwald suggestion is exactly what we would say, you should maintain parameter position. Refactoring by adding/removing an argument would result in a very hard to see bug. We maintain parameter order across APIs in the framework.

Reply:
Fair enough, I can understand that. I guess that applies to the changes in `TaskLoggingHelper`as well then? I won't mind changing this.

Reply:
I do think it has to be after `eventTimestamp` though, since `messageArgs` is a params parameter and has to come last right?

Reply:
Yes, params last.

Reply:
I've made the changes for Error now. Just want to make sure that this is the right direction before I change Warning as well. It feels weird to me, but maybe I'm just missing something ;).

Reply:
It feels weird because?

Reply:
Because something that used to be part of the messageArgs array now becomes the helpLink as far as I can tell.

Reply:
> part of the messageArgs array now becomes the helpLink as far as I can tell.

Can messageArgs start with a string? If so, wouldn't this silently change how any code with a string as the first messageArg works?

Reply:
@terrajobst can we borrow a few minutes for a micro–API-change review here?

My read on the `params` binding is that @Forgind is right to worry ([sharplab demo of binding to the with-a-`string` overload over the `params` overload](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1RhyALAQPIAUYAdsAQB5QEDKwAE4cA5gQDOfAA4BDIbIC2E5gCMAVgFMAxsADaAXQIB7CQEpyAXyo18RRqw5deBOQuVqtuwyfNWb1AG09gQAsixmFEE0zCwADGYA3NE0KdSsqHwARACexllJOGkEAPQlrOjZeVnZEsaKmjUEWW5KEgXJ2DExQdbYlkA===)). But that wouldn't be a problem with putting it at the end of the `string`s.

Reply:
Your interpretation is correct. C# rules for `params` goes something like this: if there is an applicable overload without expanding params, it will be called. I suggest keeping `eventTimestamp` last as a tie breaker:

```diff
 public BuildErrorEventArgs(/* ..., */ string senderName);
 public BuildErrorEventArgs(/* ..., */ string senderName, System.DateTime eventTimestamp);
 public BuildErrorEventArgs(/* ..., */ string senderName, System.DateTime eventTimestamp, params object[] messageArgs);
+public BuildErrorEventArgs(/* ..., */ string senderName, string helpLink, System.DateTime eventTimestamp, params object[] messageArgs);
```

Yes, this means that `eventTimestamp` changes position, but in my opinion that change isn't all that different of overloads were we keep `ref` last, for instance.

Reply:
Okay, I'll make the necessary changes some time today hopefully ;).